include/anvill/Arch.h

```diff
@@ -0,0 +1,187 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include <llvm/IR/DebugInfoMetadata.h>
+#include <llvm/IR/Function.h>
+#include <llvm/IR/Instructions.h>
+#include <llvm/IR/IntrinsicInst.h>
+#include <llvm/IR/Metadata.h>
+#include <llvm/IR/Module.h>
+#include <llvm/IR/Type.h>
+
+#include "Decl.h"
+
+namespace remill {
+class Arch;
+class IntrinsicTable;
+struct Register;
+}  // namespace remill
+
+namespace anvill {
+
+enum SizeConstraint : unsigned {
+  kBit8 = (1 << 0),
+  kBit16 = (1 << 1),
+  kBit32 = (1 << 2),
+  kBit64 = (1 << 3),
+  kBit128 = (1 << 4),
+
+  kMaxBit128 = kBit128 | kBit64 | kBit32 | kBit16 | kBit8,
+  kMaxBit64 = kBit64 | kBit32 | kBit16 | kBit8,
+  kMaxBit32 = kBit32 | kBit16 | kBit8,
+
+  kMinBit128 = kBit128,
+  kMinBit64 = kBit128 | kBit64,
+  kMinBit32 = kBit128 | kBit64 | kBit32,
+};
+
+enum TypeConstraint : unsigned {
+  kTypeInt = (1 << 0),
+  kTypePtr = (1 << 1),
+  kTypeFloat = (1 << 2),
+  kTypeVec = (1 << 3),
+
+  kTypeIntegral = kTypeInt | kTypePtr,
```

Comment:
Recommend using `llvm::DataLayout::isNonIntegralPointerType` when trying to pass a pointer, and then if it doesn't fit, have a `LOG(FATAL)` or something, and then have some kind of message related that, and file an issue related to it, and then reference the issue # in the fatal message, asking to basically tell us, via the issue, about code that fails this.